device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/AmqpsSessionHandler.java

```diff
@@ -105,7 +127,12 @@ public void onSessionFinal(Event e)
     public void onSessionRemoteOpen(Event e)
     {
         log.trace("Device session opened remotely for device {}", this.getDeviceId());
-        if (this.deviceClientConfig.getAuthenticationType() == DeviceClientConfig.AuthType.X509_CERTIFICATE)
+        this.sessionOpenedRemotely = true;
+        if (this.sessionHandlerClosedBeforeRemoteSessionOpened)
+        {
+            this.session.close();
```

Comment:
Since we're intentionally closing this session but not opening a new one in this method we should definitely log it as an error/warning. 

Do we want session.close() to throw here ?

Reply:
The connection handler that manages 1 to many of these session handlers is generally in charge of managing which sessions are opened and closed. In this particular case, there is no need to open a new session from this session handler since the connection handler deliberately closed this session handler.

The expected flow here during retry logic is to have the connection handler create a new session handler for the device that is reconnecting, and if that session handler doesn't open in time, then we close it, wait for a few seconds, and then open a new session handler to try again

Reply:
In essence, the session handler is disposable, just like the amqp session itself is here

Reply:
> Do we want session.close() to throw here ?

This is expected behavior to respond to the service acknowledging the beginning of a session that we no longer track, so no

Reply:
Got it, so what ever is calling these onXXXX methods due to the event should have context of the state and this just fired out of order due to some unforeseen circumstance.

Still probably worth logging that this condition happened either here or in the closeSession event above

Reply:
Good idea, I can add some extra logs for this